src/sync.cpp

```diff
@@ -2521,24 +2559,791 @@ void Syncs::resumeResumableSyncsOnStartup()
+  , mBackupIdToConfig()
+  , mSlot(0)
+  , mDirty(false)
+{
+}
 
+JSONSyncConfigDB::JSONSyncConfigDB(const LocalPath& dbPath)
+  : mDBPath(dbPath)
+  , mDrivePath()
+  , mBackupIdToConfig()
+  , mSlot(0)
+  , mDirty(false)
+{
+}
+
+JSONSyncConfigDB::~JSONSyncConfigDB()
+{
+    // Drop the configs.
+    clear(false);
+}
+
+const SyncConfig* JSONSyncConfigDB::add(const SyncConfig& config,
+                                        const bool flush)
+{
+    auto it = mBackupIdToConfig.find(config.getBackupId());
+
+    // Do we already have a config with this tag?
+    if (it != mBackupIdToConfig.end())
+    {
+        // Mark the database as being dirty.
+        mDirty |= flush;
+
+        // Update the config.
+        it->second = config;
+
+        // We're done.
+        return &it->second;
+    }
+
+    // Add the config to the database.
+    auto result = mBackupIdToConfig.emplace(config.getBackupId(), config);
+
+    // Mark the database as being dirty.
+    mDirty |= flush;
+
+    // We're done.
+    return &result.first->second;
+}
+
+void JSONSyncConfigDB::clear(const bool flush)
+{
+    // Are there any configs to remove?
+    if (mBackupIdToConfig.empty())
+    {
+        // Nope.
+        return;
+    }
+
+    // Mark the database as being dirty.
+    mDirty |= flush;
+
+    // Clear the config database.
+    mBackupIdToConfig.clear();
+}
+
+const JSONSyncConfigMap& JSONSyncConfigDB::configs() const
+{
+    return mBackupIdToConfig;
+}
+
+bool JSONSyncConfigDB::dirty() const
+{
+    return mDirty;
+}
+
+const LocalPath& JSONSyncConfigDB::dbPath() const
+{
+    return mDBPath;
+}
+
+const LocalPath& JSONSyncConfigDB::drivePath() const
+{
+    return mDrivePath;
+}
+
+const SyncConfig* JSONSyncConfigDB::getByBackupId(handle backupId) const
+{
+    auto it = mBackupIdToConfig.find(backupId);
+
+    if (it != mBackupIdToConfig.end())
+    {
+        return &it->second;
+    }
+
+    return nullptr;
+}
+
+const SyncConfig* JSONSyncConfigDB::getByRootHandle(handle targetHandle) const
+{
+    if (targetHandle == UNDEF)
+    {
+        return nullptr;
+    }
+
+    for (auto& it : mBackupIdToConfig)
+    {
+        if (it.second.getRemoteNode() == targetHandle)
+        {
+            return &it.second;
+        }
+    }
+
+    return nullptr;
+}
+
+error JSONSyncConfigDB::read(JSONSyncConfigIOContext& ioContext)
+{
+    vector<unsigned int> confSlots;
+
+    // Determine which slots we should load first, if any.
+    if (ioContext.getSlotsInOrder(mDBPath, confSlots) != API_OK)
+    {
+        // Couldn't get a list of slots.
+        return API_ENOENT;
+    }
+
+    // Try and load the database from one of the slots.
+    for (const auto& slot : confSlots)
+    {
+        // Can we read the database from this slot?
+        if (read(ioContext, slot) == API_OK)
+        {
+            // Update the slot number.
+            mSlot = (slot + 1) % NUM_SLOTS;
+
+            // Yep, loaded.
+            return API_OK;
+        }
```

Comment:
I've improved the logging of DB deserialization, reading and writing.

I have not added an assertion here, however, as I need to be able to test the "try the next slot" logic.

Hopefully this will be sufficient.